mesonbuild/interpreterbase/decorators.py

```diff
@@ -512,23 +524,20 @@ def wrapper(*wrapped_args: T.Any, **wrapped_kwargs: T.Any) -> T.Any:
 
             def emit_feature_change(values: T.Dict[_T, T.Union[str, T.Tuple[str, str]]], feature: T.Union[T.Type['FeatureDeprecated'], T.Type['FeatureNew']]) -> None:
                 for n, version in values.items():
-                    warn = False
                     if isinstance(version, tuple):
                         version, msg = version
                     else:
                         msg = None
 
-                    if n in {dict, list}:
-                        assert isinstance(n, type), 'for mypy'
+                    if isinstance(n, ContainerTypeInfo):
+                        if n.check_any(value):
+                            feature.single_use(f'"{name}" keyword argument "{info.name}" of type "{n.description()}"', version, subproject, msg, location=node)
+                    elif isinstance(n, type):
                         if isinstance(value, n):
                             feature.single_use(f'"{name}" keyword argument "{info.name}" of type {n.__name__}', version, subproject, msg, location=node)
-                    elif isinstance(value, (dict, list)):
-                        warn = n in value
                     else:
-                        warn = n == value
-
-                    if warn:
-                        feature.single_use(f'"{name}" keyword argument "{info.name}" value "{n}"', version, subproject, msg, location=node)
+                        if n == value:
+                            feature.single_use(f'"{name}" keyword argument "{info.name}" value "{n}"', version, subproject, msg, location=node)
```

Comment:
There's a lot of repetition here in the `feature.single_use(...¸version, subproject, msg, location=node)`

I think this could be simplified by doing something like:
```python
warning: T.Optional[str] = None
if ...
endif
if warning is not None:
    feature.single_use(warning, version, subproject, msg, location=node)
```

Reply:
done